### PR TITLE
feat: add interestingness and soft constraints

### PR DESCRIPTION
**Pull request by JunranY**

### CODE REVIEW

File: docs/applications/draco_learn.ipynb [hunk 1]
```diff
@@ -362,13 +362,22 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
+<<<<<<< HEAD
```
Comment (domoritz) [inline docs/applications/draco_learn.ipynb:365]: Merge conflict markers

**Review by domoritz — CHANGES_REQUESTED** [review]

File: draco/asp/soft.lp [hunk 2]
```diff
@@ -268,31 +279,35 @@ preference(color_entropy_high,E) :-
     attribute((encoding,field),E,F),
     attribute((encoding,channel),E,color),
     helper((encoding,scale_type),E,linear),
-    EN > 5000.
+    helper(enc_interesting,E),
+    EN > 12000.
```
Comment (domoritz) [inline draco/asp/soft.lp:283]: Why did you change these values?

File: draco/asp/soft.lp [hunk 3]
```diff
@@ -626,145 +641,214 @@ preference(summary_rect,M) :-
...
+    helper(mark_encoding_discrete_or_binned,_,E),
+    helper(enc_interesting,E).
 
 % @soft(summary_discrete_text) Discrete text for summary tasks.
 preference(summary_discrete_text,E) :-
     attribute(task,root,summary),
     attribute((encoding,channel),E,text),
-    helper(mark_encoding_discrete_or_binned,_,E).
+    helper(mark_encoding_discrete_or_binned,_,E),
+    helper(enc_interesting,E).
```
Comment (domoritz) [inline draco/asp/soft.lp:799]: It feels strange to me to add interestingness to all these preferences. Won't that break if we don't have interestingness?

File: draco/tests/test_soft.py [hunk 4]
```diff
@@ -5385,11 +5395,13 @@ def test_value_continuous_x():
         list_preferences(
             b.program
             + """
+    attribute((field,interesting),precipitation,true).
```
Comment (domoritz) [inline draco/tests/test_soft.py:5398]: This seems wrong to me. I don't think we want to have people have to add interestedness to all fields just to make the recommendations work.

**Review by domoritz — CHANGES_REQUESTED** [review]
I don't think we should require interestingness for all the fields. Also, please add docs.

File: draco/asp/soft.lp (hunk 2 above)
Comment (JunranY) [reply draco/asp/soft.lp:283]: the updated one is the same as in draco1. The previous values was written by @Zehua-Zeng and the reason I remembered was that she observed in the kim2018 dataset, the entropy values never exceeded 12000? I changed it back to check the testing score for draco-learn.

File: draco/asp/soft.lp (hunk 3 above)
Comment (JunranY) [reply draco/asp/soft.lp:799]: with the changes, these presences surely have different meaning now. However I also checked that the current version is what meant by draco1 ("for summary task, the interested column are preferred to be discrete text. I also changed the test cases for them to not be broken.

File: draco/tests/test_soft.py (hunk 4 above)
Comment (JunranY) [reply draco/tests/test_soft.py:5398]: As I interpreted from the kim2018 dataset and their paper, the interestingness is associated with the tasks. something like "identifying the values for some field". So I think if people don't specify the task or interestingness, the recommendation would still work, but probably generate less precise results.

File: draco/asp/soft.lp (hunk 2 above)
Comment (domoritz) [reply draco/asp/soft.lp:283]: Okay. I don't think the prior values were super thought out so your values may be better.

File: draco/asp/soft.lp (hunk 3 above)
Comment (domoritz) [reply draco/asp/soft.lp:799]: I think we should have simpler constraints when possible. Would it hurt to remove the restriction to interesting fields from the constraints?

File: draco/tests/test_soft.py (hunk 4 above)
Comment (domoritz) [reply draco/tests/test_soft.py:5398]: I see. Yes, I think it makes sense to say tasks need associated interesting fields.

Comment (JunranY) [reply draco/tests/test_soft.py:5398]: In addition there are soft constraints about only task and mark (without interestingness).

File: draco/asp/soft.lp (hunk 3 above)
Comment (domoritz) [reply draco/asp/soft.lp:799]: I think we should have simpler constraints when possible. Do we use interestingness only with task constraints? Would it hurt to remove the restriction to interesting fields from constraints that don't involve tasks?

File: docs/facts/schema.md [hunk 5]
```diff
@@ -39,6 +39,9 @@ the actual entropy).
 
 `(field,freq)` : The frequency of the most common value. Only used for strings.
 
+`(field,interesting)` : When the [`task`](task.md) regards specific fields, fields can be marked as relevant to the task
+with interesting.
```
Comment (domoritz) [inline docs/facts/schema.md:43]:
```suggestion
`(field,interesting)` : When the [`task`](task.md) regards specific fields, fields can be marked as relevant to the task.
```

**Review by domoritz — APPROVED** [review]
